File: src/my_py_pkg/my_py_pkg/first_node_oop.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node


class PY_NODE(Node):

    def __init__(self):
        super().__init__("py_test")
        self.counter_ = 0
        self.get_logger().info("Hello World")
        self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        self.get_logger().info("HELLO " + str(self.counter_))
        self.counter_ += 1


def main(args=None):
    rclpy.init(args=args)
    node = PY_NODE()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()


# import rclpy
# from rclpy.node import Node


# class MyNode(Node):
#     def __init__(self):
#         super().__init__("node")
#         self.get_logger().info("Hello World")
#         self.create_timer(1.0, self.timer_callback)

#     def timer_callback(self):
#         self.get_logger().info("Hello")


# def main(args=None):
#     rclpy.init(args=args)
#     node = MyNode()
#     rclpy.spin(node)
#     rclpy.shutdown()


# if __name__ == "__main__":
#     main()

File: src/my_cpp_pkg/src/first_node_oop.cpp
#include "rclcpp/rclcpp.hpp"

class NewNode : public rclcpp::Node {
public:
  NewNode() : Node("cpp_test_oop") {
    RCLCPP_INFO(this->get_logger(), "Hello World");
    counter_ = 0;
    timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                     std::bind(&NewNode::timer_callback, this));
  }

private:
  void timer_callback() {
    RCLCPP_INFO(this->get_logger(), "Hello %d", counter_);
    counter_ += 1;
  }
  rclcpp::TimerBase::SharedPtr timer_;
  int counter_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NewNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

// #include "rclcpp/rclcpp.hpp"

// class MyNode : public rclcpp::Node {
// public:
//   MyNode() : rclcpp::Node("New_node") {
//     RCLCPP_INFO(get_logger(), "Hello World");
//     timer_ = this->create_wall_timer(std::chrono::seconds(1),
//                                      std::bind(&MyNode::timerCallback,
//                                      this));
//   }

// private:
//   void timerCallback() { RCLCPP_INFO(get_logger(), "Hello"); }
//   rclcpp::TimerBase::SharedPtr timer_;
// };

// int main(int argc, char **argv) {
//   rclcpp::init(argc, argv);
//   auto node = std::make_shared<MyNode>();
//   rclcpp::spin(node);
//   rclcpp::shutdown();
//   return 0;
// }
